File: python/2020/day08.py
import re
from common2020 import str_list, print_ans


class Instruction:

    def __init__(self, operation, argument):
        self.operation = operation
        self.argument = argument
        self.read = False


def read_instructions() -> list:
    input_list = str_list(8)
    instructions = [construct_instruction(line) for line in input_list]

    return instructions


def construct_instruction(line) -> Instruction:
    inst_re = re.compile(r'(^\w\w\w)(?:\s)([+|-]\d+)')
    inst_match = inst_re.match(line)

    operation = inst_match.group(1)
    argument = int(inst_match.group(2))

    return Instruction(operation, argument)


def reset():
    for instruction in instructions:
        instruction.read = False 


def is_acc(operation): 
    return operation == "acc"


def is_jmp(operation): 
    return operation == "jmp"


def is_nop(operation): 
    return operation == "nop"


def flip(operation):
    if is_jmp(operation):
        return "nop"
    if is_nop(operation):
        return "jmp"
    return operation


def run():
    accumulator = 0
    cursor_index = 0
    max_index = len(instructions)

    while cursor_index < max_index:
        if instructions[cursor_index].read:
            return accumulator, False

        operation = instructions[cursor_index].operation
        argument = instructions[cursor_index].argument

        instructions[cursor_index].read = True

        if is_acc(operation):
            accumulator += argument
        elif is_jmp(operation):
            cursor_index += argument
            continue

        cursor_index += 1

    return accumulator, True


def pt_1():
    return run()[0]


def pt_2():
    for instruction in instructions:
        reset()

        # Skip acc lines
        if is_acc(instruction.operation):
            continue

        # Flip instruction
        instruction.operation = flip(instruction.operation)

        # Test boot
        accumulator, boot_sucessful = run()

        if boot_sucessful: 
            return accumulator

        # Reset instruction if fails to boot
        instruction.operation = flip(instruction.operation)

    return accumulator


def main():
    pt1 = pt_1()
    pt2 = pt_2()
    print_ans(pt1, pt2)
    

instructions = read_instructions()
main()

File: solutions/typescript/2020/day08.ts
import { input, print } from "common";

type instruction = {
  operation: string,
  argument: number
}

type bootLog = {
  accumulator: number,
  terminates: boolean
}

function instructionList(raw: string): instruction[] {
  return raw
    .split("\n")
    .map((line: string) => formatInstruction(line));
}

function formatInstruction(rawInst: string): instruction {
  const inst = rawInst.split(" ");
  const op = inst[0];
  const arg = Number.parseInt(inst[1]);
  return {
    operation: op,
    argument: arg
  };
}

function flipInstruction(inst: instruction): instruction {
  let op = inst.operation;
  const arg = inst.argument;

  if (op === "jmp") {
    op = "nop";
  }
  else if (op === "nop") {
    op = "jmp";
  }
  return {
    operation: op,
    argument: arg
  };
}

function boot(instructions: instruction[]): bootLog {
  let accumulator = 0;
  let cursorIndex = 0;
  let terminates = true;
  const cache: Record<number, instruction> = {};
  const maxIndex = instructions.length;

  while (cursorIndex < maxIndex) {
    if (cache[cursorIndex] !== undefined) {
      terminates = false;
      break;
    }

    cache[cursorIndex] = instructions[cursorIndex];

    if (cache[cursorIndex].operation === "acc") {
      accumulator += cache[cursorIndex].argument;
      cursorIndex += 1;
    }
    else if (cache[cursorIndex].operation === "jmp") {
      cursorIndex += cache[cursorIndex].argument;
    }
    else { // "nop"
      cursorIndex += 1;
    }
  }

  return {
    accumulator: accumulator,
    terminates: terminates
  };
}

function pt1(instructions: instruction[]): number {
  return boot(instructions).accumulator;
}

function pt2(instructions: instruction[]): number {
  for (let i = 0; i < instructions.length; i++) {
    if (instructions[i].operation === "acc") {
      continue;
    }

    instructions[i] = flipInstruction(instructions[i]);
    const bootLog = boot(instructions);

    if (bootLog.terminates) {
      return bootLog.accumulator;
    }
    instructions[i] = flipInstruction(instructions[i]);
  }
  return -1;
}

const instructions = instructionList(input(2020, 8));
print(pt1(instructions), pt2(instructions));
